tests/unit/test_middleware/test_constraints.py

```diff
@@ -0,0 +1,365 @@
+from typing import Any
+
+import pytest
+
+from litestar.middleware.base import ASGIMiddleware
+from litestar.middleware.constraints import (
+    ConstraintViolationError,
+    CycleError,
+    MiddlewareConstraints,
+    check_middleware_constraints,
+)
+from litestar.types import ASGIApp, Receive, Scope, Send
+
+
+def fully_qualified_name(obj: Any) -> str:
+    return f"{obj.__module__}.{obj.__qualname__}"
+
+
+# place these here so they are importable
+
+
+class MiddlewareOne(ASGIMiddleware):
+    async def handle(self, scope: Scope, receive: Receive, send: Send, next_app: ASGIApp) -> None:
+        pass
+
+
+class MiddlewareTwo(ASGIMiddleware):
+    async def handle(self, scope: Scope, receive: Receive, send: Send, next_app: ASGIApp) -> None:
+        pass
+
+
+class MiddlewareThree(ASGIMiddleware):
+    async def handle(self, scope: Scope, receive: Receive, send: Send, next_app: ASGIApp) -> None:
+        pass
+
+
+def middleware_factory(app: ASGIApp) -> ASGIApp:
+    async def wrapped_app(scope: Scope, receive: Receive, send: Send) -> None:
+        pass
+
+    return wrapped_app
+
+
+def test_constraint_ignore_not_found() -> None:
+    constraints = (
+        MiddlewareConstraints()
+        .apply_after("SomethingNotAvailable", ignore_not_found=True)
+        .apply_before("SomethingNotAvailable", ignore_not_found=True)
+    )
+    resolved = constraints.resolve()
+    assert resolved.after == ()
+    assert resolved.before == ()
+
+
+def test_order_before() -> None:
+    one = MiddlewareOne()
+    two = MiddlewareTwo()
+    three = MiddlewareThree()
+
+    two.constraints = MiddlewareConstraints(before=(MiddlewareOne,))
+
+    middlewares = (
+        one,
+        two,
+        three,
+    )
+
+    with pytest.raises(ConstraintViolationError, match="MiddlewareTwo.*before.*MiddlewareOne"):
+        check_middleware_constraints(middlewares)
+
+
+def test_order_after() -> None:
+    one = MiddlewareOne()
+    two = MiddlewareTwo()
+    three = MiddlewareThree()
+
+    two.constraints = MiddlewareConstraints(after=(MiddlewareOne,))
+
+    middlewares = (two, one, three)
+
+    with pytest.raises(ConstraintViolationError, match="MiddlewareTwo.*after.*MiddlewareOne"):
+        check_middleware_constraints(middlewares)
+
+
+def test_order_string_ref() -> None:
+    one = MiddlewareOne()
+    two = MiddlewareTwo()
+    three = MiddlewareThree()
+
+    two.constraints = MiddlewareConstraints().apply_before(fully_qualified_name(MiddlewareOne))
```

Comment:
im not too sure I get the meaning of apply_before vs before.
```
two.constraints = MiddlewareConstraints(before=(MiddlewareOne,))
two.constraints = MiddlewareConstraints().apply_before(fully_qualified_name(MiddlewareOne))
```

Reply:
It's just a convenience method that allows other options, such as passing in string references and `ignore_not_found`

Reply:
mmm ok that's smart